Validate nested data, add error details (#151)
yiisoft/validator: README.md

Table of Contents
=================

* [Yii Validator](#yii-validator)
    * [Features](#features)
    * [Requirements](#requirements)
    * [Installation](#installation)
    * [General usage](#general-usage)
        * [Validating a single value](#validating-a-single-value)
        * [Validating a set of data](#validating-a-set-of-data)
            * [Skipping validation on error](#skipping-validation-on-error)
            * [Skipping empty values](#skipping-empty-values)
            * [Nested and related data](#nested-and-related-data)
        * [Conditional validation](#conditional-validation)
        * [Creating your own validation rules](#creating-your-own-validation-rules)
        * [Grouping multiple validation rules](#grouping-multiple-validation-rules)
        * [Setting up your own formatter](#setting-up-your-own-formatter)
    * [Testing](#testing)
        * [Unit testing](#unit-testing)
        * [Mutation testing](#mutation-testing)
        * [Static analysis](#static-analysis)
    * [License](#license)
    * [Support the project](#support-the-project)
    * [Follow updates](#follow-updates)

> Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)

Reviewer: I don't think we need it in README.

Author: Sorry, not 100% clear for me.

You mean the whole TOC? Why not? It's getting quite big and becoming harder to navigate. For better sync we can automate it. 

Or you discourage using this specific tool? 

Or mentioning it in `README` (so you mean only the line "Created by `gh-md-toc`..." is redundant)? But I think it's useful - it marks this block as auto generated (we can also add more explicit note) and it contains link to instructions on how to update it.

Reviewer: We are eventually going to render these at the website like https://www.yiiframework.com/extension/yiisoft/yii2-redis. There we generate TOC automatically:

<img width="612" alt="image" src="https://user-images.githubusercontent.com/47294/151163327-07b2d34e-bd09-4c4b-8da7-f31b4733438a.png">

"Created by gh-md-toc" despite being technically useful is very discouraging for the user of the library.

Author: > We are eventually going to render these at the website

But it's not ready yet. And I think there are still many users reading docs directly from Github, but I don't have stats to prove it.

Anyway, I got it. Removed the whole block.

Author: :heavy_check_mark:

            'points' => [
                ['coordinates' => ['x' => -1, 'y' => 1], 'rgb' => [0, 0, 0]],
                ['coordinates' => ['x' => -2, 'y' => 2], 'rgb' => [255, 255, 255]],
            ],
        ],
        [
            'points' => [
                ['coordinates' => ['x' => -13, 'y' => 13], 'rgb' => [-3, 258, 0]],
                ['coordinates' => ['x' => -14, 'y' => 14], 'rgb' => [0, -4, 259]],
            ],
        ],
    ],
];
$rule = Nested::rule([
    'charts' => Each::rule(new Rules([
        Nested::rule([
            'points' => Each::rule(new Rules([
                Nested::rule([
                    'coordinates' => Nested::rule([
                        'x' => [Number::rule()->min(-10)->max(10)],
                        'y' => [Number::rule()->min(-10)->max(10)],
                    ]),
                    'rgb' => Each::rule(new Rules([
                        Number::rule()->min(0)->max(255)->skipOnError(false),
                    ])),
                ])->skipOnError(false),
            ])),
        ])->skipOnError(false),
    ])),
]);
$errors = $rule->validate($data)->getErrors();
```

The contents of the errors will be:

```php
$errors = [

>     'charts.0.points.0.coordinates.x' => 'Value must be no less than -10.',
>     'charts.0.points.0.rgb.0' => 'Value must be no less than 0. -1 given.',

Reviewer 2: ```suggestion
    'Value must be no less than -10.',
    'Value must be no less than 0. -1 given.',
```

Author: Thanks, but I think in this context it's better to change the called method.

Author: :heavy_check_mark:

Number::rule()->integer()->max(100)->skipOnEmpty(true)
```

> #### Nested and related data
> 

Reviewer: ```suggestion
#### Nested and related data

In many cases there is a need to validate data consisting of many similar structures. There is a `Nested` rule for this purpose:
```

Author: Good addition. But I think I'll correct this a bit. `Nested` is more about relations, while `Each` - about similar structures.

Author: :heavy_check_mark:

```php
use Yiisoft\Validator\Rule\HasLength;
use Yiisoft\Validator\Rule\Nested;
use Yiisoft\Validator\Rule\Number;
use Yiisoft\Validator\Rule\Required;

$data = ['author' => ['name' => 'Alexey', 'age' => '31']];
$rule = Nested::rule([
    'author' => Nested::rule([
        'name' => [HasLength::rule()->min(3)],
        'age' => [Number::rule()->min(18)],
    )];
]);
$errors = $rule->validate($data)->getErrors();
```

> A more complex real-life example using arrays.

Reviewer: ```suggestion
A more complex real-life example is a chart that is made of points. This data is represented as an arrays:
```

Author: Thanks. Added with small correction and addition. See https://github.com/yiisoft/validator/pull/151#discussion_r793256574. Also fixed a grammar error.

Author: :heavy_check_mark:

            ],
        ],
        [
            'dots' => [
                ['coordinates' => ['x' => -1, 'y' => 1], 'rgb' => [0, 0, 0]],
                ['coordinates' => ['x' => -2, 'y' => 2], 'rgb' => [255, 255, 255]],
            ],
        ],
        [
            'dots' => [
                ['coordinates' => ['x' => -13, 'y' => 13], 'rgb' => [-3, 258, 0]],
                ['coordinates' => ['x' => -14, 'y' => 14], 'rgb' => [0, -4, 259]],
            ],
        ],
    ],
];
$rule = Nested::rule([
    'charts' => Each::rule(new Rules([
        Nested::rule([
            'dots' => Each::rule(new Rules([
                Nested::rule([
                    'coordinates' => Nested::rule([
                        'x' => [Number::rule()->min(-10)->max(10)],
                        'y' => [Number::rule()->min(-10)->max(10)],
                    ]),
                    'rgb' => Each::rule(new Rules([
                        Number::rule()->min(0)->max(255)->skipOnError(false),
                    ])),
                ])->skipOnError(false),
            ])),
        ])->skipOnError(false),
    ])),
]);
$errors = $rule->validate($data)->getErrors();
```

> The contents of the errors will be (omitted for brevity):

Reviewer: ```suggestion
The contents of the errors will be:
```

Author: :heavy_check_mark:

```php
$errors = [
    'charts.0.dots.0.coordinates.x' => 'Value must be no less than -10.',

>     'charts.0.dots.0.rgb.0' => 'Value must be no less than 0. -1 given.',
> ];

Reviewer: ```suggestion
    'charts.0.dots.0.rgb.0' => 'Value must be no less than 0. -1 given.',
    // ...
];
```

Author: :heavy_check_mark:
